### PR TITLE
Serialization refactor step 2

### PR DESCRIPTION
**Pull request by greg-szabo**
Implemented null-value deserialization for Vec<u8> using Visitor.

This is part 2 of the serialization improvement efforts described in #247 .

(This PR was opened after part 1 was merged. It was easier to open this than to keep the original #257)

This PR is ready for review.

* [X] Referenced an issue explaining the need for the change
* [ ] Updated all relevant documentation in docs
* [X] Updated all code comments where relevant
* [X] Wrote tests
* [X] Updated CHANGES.md

### CODE REVIEW

**Comment by greg-szabo** [comment]
I think this is more to do about the strictness in Rust. When you omit a value in JSON, it defaults to `null`. Rust can deal with omitted values by using the `Default` trait. But if you explicitly set a value to `null` in JSON, that's not strictly a default value in Rust (becuase you can set the default to be, say "42"). Hence it's a value you need to receive and process. By default, `null`s are `None`s in Rust, so with an `Option<>` we would be able to receive and process it. But the whole point for us was to get away from Options, to make life easier for the developer, hence we need a "layer" of code that receives the explicit `null` value and turns it into something else. That's where the Visitor comes in.
If serde wanted to deal with this internally, they would need to assume that a `null` value translates to the default value. Since, even the `Default` trait is optional, I think they wanted to keep this optional too. Maybe an annotation for null values would help, like there is one for default values.

**Comment by Liamsi** [comment]
I might have thought that serde does treat `null` as an empty vector. Apparently it doesn't: https://github.com/serde-rs/serde/issues/836

So the code looks fine to me, I was just wondering if we can do without the extra visitor here. Also I though we'd have a separate serilziation type (which could have an option in the fields and the actual ideal tendermint type that just has a vec and From impls between those types. 

Maybe I'm overthinking all this  😄

**Comment by greg-szabo** [comment]
In the case of `Vec<u8>` it doesn't seem to make sense to turn it into a custom type (which would possibly allow us to use `Option<>` and convert internally, like in CommitSig) because the custom type wouldn't add any value (HexBytes and Base64Bytes are only different on the JSON side, it doesn't make any difference for the developer).
Built-in types can only be extended in the library we use them in (everywhere one-by-one) which would sprinkle serialization code all over the place. Or we write an annotation (like in the PR), which will require the Visitor so it can convert between two different types.

So, with annotating built-in types from weird formats, we might need to use a Visitor (like in this case). For custom types, most probably we won't need it, we'll have an internal struct (with `Option<>` or whatever it needs) and convert from that.

At least that's how I see it. If you think we can do this without the Visitor, point me in the right direction.

**Comment by melekes** [comment]
> I might have thought that serde does treat null as an empty vector. 

Maybe the solution would be to not have explicit `null`s?

**Comment by Liamsi** [comment]
> Maybe the solution would be to not have explicit nulls? Is it possible?

I would like that. But it would require some changes in the go code too and we'd need to communicate this clearly in the spec (nulls are not uncommon in JSON). 
The work in https://github.com/tendermint/tendermint/issues/4828 could be an opportunity to validate this idea.

**Comment by Liamsi** [comment]
> In the case of Vec<u8> it doesn't seem to make sense to turn it into a custom type

I didn't mean Vec<u8> type but the wrapping struct which has `Vec<u8>` as a field. E.g. let's say the Header:

`serilaization::block::Header <---- TryFrom ----> core::block::Header`

The left header can use an Option in the fields or whatever is necessary to correctly deserialize this and the TryFrom handles this accordingly and translates into a real, core value on the right (with absolutely no serde annotation whatsoever). See https://github.com/informalsystems/tendermint-rs/issues/197#issuecomment-606103374

Downside would be a lot of duplication as the types would be almost identical. But it would be really, really simple, too. Also, there would be a cleat separation between serialization types and ideal core types that hold more logic and with which devs would mainly deal with. Devs wouldn't need to care at all about the serialization specifics. The only thing they need to now is that under a `serialization` module, there are types from which you can use to (de)serilalize and translate to the core types. Does al that make any sense?

That all said, this PR is an improvement to current situation and excellent work! The above rather touches upon that we didn't yet decide on a clear strategy on serialization as a whole (and not about specific core types).I think we should merge this PR but keep in mind that there is a bigger question on howto handle serialization (also of different serialization kinds, e.g. binary proto/amino and different kinds of JSON). If we decide to go with the above mentioned approach, some of the custom serialization code (the questioned Visitor), can be deleted.

File: tendermint/src/serializers.rs [hunk 1]
```diff
@@ -96,24 +97,83 @@ pub mod time_duration {
...
             D: Deserializer<'de>,
         {
-            let string = String::deserialize(deserializer)?;
-
-            hex::decode_upper(&string)
-                .or_else(|_| hex::decode(&string))
-                .map_err(deError::custom)
+            deserializer.deserialize_option(super::BytesVisitor {
+                string_type: super::ByteStringType::Hex,
+            })
```
Comment (Liamsi) [inline tendermint/src/serializers.rs:176]: So this `deserialize_option` here and for the "ByteTypes" introduced the need for a Visitor, right? 

No need to apply this but if replaced by sth along the lines of:

```suggestion
let string = Option::<String>::deserialize(deserializer)?.unwrap_or("".into());
            hex::decode_upper(&string)
                .or_else(|_| hex::decode(&string))
                .map_err(serde::de::Error::custom)
```
here and in the other two places and if I'm not wrong the need for a custom Visitor would be gone?

Comment (Liamsi) [reply tendermint/src/serializers.rs:176]: Confirmed: https://github.com/informalsystems/tendermint-rs/pull/271/files

**Pull request merged**